dali/imgcodec/decoders/decoder_test_helper.h

```diff
@@ -39,26 +40,96 @@ namespace test {
-template<typename OutputType>
-class CpuDecoderTestBase : public ::testing::Test {
+template<typename Backend, typename OutputType>
+class DecoderTestBase : public ::testing::Test {
  public:
-  CpuDecoderTestBase() : tp_(4, CPU_ONLY_DEVICE_ID, false, "Decoder test") {}
+  explicit DecoderTestBase(int threads_cnt = 4)
+    : tp_(threads_cnt, GetDeviceId(), false, "Decoder test") {}
 
   /**
-  * @brief Decodes an image and returns the result as a tensor.
+  * @brief Decodes an image and returns the result as a CPU tensor.
   */
   Tensor<CPUBackend> Decode(ImageSource *src, const DecodeParams &opts = {}, const ROI &roi = {}) {
     EXPECT_TRUE(Parser()->CanParse(src));
     ImageInfo info = Parser()->GetInfo(src);
 
     Tensor<CPUBackend> result;
     EXPECT_TRUE(Decoder()->CanDecode(src, opts));
-    TensorShape<> shape = (roi.use_roi() ? roi.shape() : info.shape);
+    auto shape = AdjustToRoi(info.shape, roi);
     result.Resize(shape, type2id<OutputType>::value);
 
-    SampleView<CPUBackend> view(result.raw_mutable_data(), result.shape(), result.type());
-    DecodeResult decode_result = Decoder()->Decode(view, src, opts, roi);
-    EXPECT_TRUE(decode_result.success);
+    if (GetDeviceId() == CPU_ONLY_DEVICE_ID) {
+      SampleView<CPUBackend> view(result.raw_mutable_data(), result.shape(), result.type());
+      DecodeResult decode_result = Decoder()->Decode(view, src, opts, roi);
+      EXPECT_TRUE(decode_result.success);
+    } else {  // GPU
+      TensorList<GPUBackend> gpu_list;
+      gpu_list.Resize(TensorListShape{{result.shape()}}, result.type());
+      SampleView<GPUBackend> view(gpu_list.raw_mutable_tensor(0), result.shape(), result.type());
+
+      auto stream = CUDAStreamPool::instance().Get(GetDeviceId());
+      auto decode_result = Decoder()->Decode(stream, view, src, opts, roi);
+      EXPECT_TRUE(decode_result.success);
+
+      TensorList<GPUBackend> cpu_list;
+      cpu_list.Copy(gpu_list);
+      result.Copy(cpu_list, 0);
+    }
+
+    return result;
+  }
+
+  /**
+   * @brief Decodes a batch of images, invoking the batch version of ImageDecoder::Decode
+   */
+  std::vector<Tensor<CPUBackend>> Decode(cspan<ImageSource *> in,
+                                         const DecodeParams &opts = {},
+                                         cspan<ROI> rois = {}) {
+    auto parser = Parser();
+    auto decoder = Decoder();
+
+    int n = in.size();
+    std::vector<TensorShape<>> shape(n);
+    std::vector<Tensor<CPUBackend>> result(n);
+    std::vector<ROI> empty_rois(n);
+
+    if (rois.size() == 0)
+      rois = make_span(empty_rois);
+    assert(rois.size() == n);
+
+    for (int i = 0; i < n; i++) {
+      EXPECT_TRUE(parser->CanParse(in[i]));
+      EXPECT_TRUE(decoder->CanDecode(in[i], opts));
+      ImageInfo info = parser->GetInfo(in[i]);
+      shape[i] = AdjustToRoi(info.shape, rois[i]);
+      result[i].Resize(shape[i], type2id<OutputType>::value);
+    }
+
+    if (GetDeviceId() == CPU_ONLY_DEVICE_ID) {
```

Comment:
Similarly to what I suggested above:
```kernels::TestTensorList<OutputType, 3> output_;  // a member

 output_.reshape(shape);  // make it a TensorListShape instead of vector<TensorShape>
  
 // For CPU
 auto tlv = output_.cpu();
 vector<SampleView<...>> samples(n);
 for (int i=0; i < n; i++) {
   auto tv = tlv[i];
   samples[i] = {tv.data, tv.shape, type2id<OutputType>::value};
 }
 // ... decode
 return tlv;  // you will return TensorListView<StorageCPU, OutputType, 3> instead
 

// For GPU
 auto tlv = output_.gpu();
 vector<SampleView<...>> samples(n);
 for (int i=0; i < n; i++) {
   auto tv = tlv[i];
   samples[i] = {tv.data, tv.shape, type2id<OutputType>::value};
 }
 // ... run decoding
 // wait for async work to finish (stream sync)
 auto tlv_cpu = output_.cpu();  // data is already copied to the CPU
 return tlv_cpu;  // return TensorListView<StorageCPU, OutputType, 3>